### PR TITLE
feat: add overview by ci-builds

### LINKED ISSUE #362: Dashboard overview by build id
- [X] I have searched issues for similar feature requests

## What is this feature about (1 sentence)
As a monorepo developer,
I want to have an overview of all affected projects by looking for a specific build id.

## Why is it needed? What is the value? For whom do we build it?
Especially when you deal with a monorepo, you have many applications getting tested in the same CI run,
you need to know if everything went smoothly or not before merging. By supporting that, it gives a much quicker overview instead of having to look into each project individually.

## High-level feature overview
Basically just a link like `/builds/<build id>/` listing all projects, runs and statuses using the build id. We can then drilldown more from there.

### PR DESCRIPTION
**Pull request by mathpaquette**
closes #362

For bug fixes:

- [X] Add a reference to the original issue
- [X] Add a test and / or some kind of instructions to verify the fix

---

For new features:

- [ ] It's better to first discuss features proposal in [Discussions](https://github.com/sorry-cypress/sorry-cypress/discussions) or [Slack](https://sorry-cypress.slack.com/join/shared_invite/zt-eis1h6jl-tJELaD7q9UGEhMP8WHJOaw#/)
- [ ] Describe the use-case in details
- [ ] Ensure the solution is backwards-compatible
- [ ] Test it. Submit screenshots / outputs / tests together with the PR.

--- Side note ---

Your contribution is very valuable, we **really appreciate your time and effort**!

Please note:

- Adding a new code and / or feature complicates the maintenance
- Most people disappear after contributing a feature, leaving the support to core maintainers

Given that, please don't be offended if your PR is rejected or significantly modified.

### CODE REVIEW

**Comment by mathpaquette** [comment]
not 100% done yet, need to polish stuff here and there but the functionality is working.
I'd like to gather some initial impressions.
image

**Comment by mathpaquette** [comment]
highlights:
- Landing page / is now redirecting automatically to /projects by default. To avoid this redirection, just need to pass redirect=false This is giving us some time deciding on a new landing page/dashboard with some useful stats.
- Created /ci-builds route
- Created a seed script in order to test the new functionality
- Broke down sidemenu into few components to reduce size
- Improved the searchField by completely removing query param when clearing search

**Comment by mathpaquette** [comment]
image
 
projects are now getting colored by severity as shown by the screenshot

**Comment by mathpaquette** [comment]
@agoldis what's missing on my side is basically an improved version of the landing page and icons in breadcrumb.
let me know.

**Comment by agoldis** [comment]
@mathpaquette Quite a big PR, will try to review during the upcoming week

**Comment by mathpaquette** [comment]
@agoldis no problem. In the meantime, do you think we can create a next branch and create a release out of it ?
I dont think how much you reviewed it already but I can try to split into more commits to facilitate the review a little bit.

File: packages/dashboard/src/components/layout/sidebar.tsx [hunk 1]
```diff
@@ -147,165 +148,11 @@ const DrawerContentContainer = styled('div', {
   overflowX: 'hidden',
 }));
 
-export const ProjectListMenu: ProjectListMenuType = ({
```
Comment (mathpaquette) [inline packages/dashboard/src/components/layout/sidebar.tsx:150]: moved into projectListMenu.tsx

File: packages/dashboard/src/components/layout/sidebar.tsx [hunk 2]
```diff
@@ -147,165 +148,11 @@ const DrawerContentContainer = styled('div', {
...
-              </IconButton>
-            </div>
-          </Tooltip>
-        );
-      })}
-    </div>
-  );
-};
-
-export const ProjectDetailsMenu: ProjectDetailsMenuType = ({
```
Comment (mathpaquette) [inline packages/dashboard/src/components/layout/sidebar.tsx:210]: moved into projectDetailsMenu.tsx

File: packages/dashboard/src/components/layout/sidebar.tsx [hunk 3]
```diff
@@ -642,19 +499,3 @@ export const Sidebar: SidebarType = ({ open, onToggleSidebar }) => {
 
 type SidebarProps = { open: boolean; onToggleSidebar?: () => void };
 type SidebarType = React.FC<SidebarProps>;
-
-type ProjectDetailsMenuProps = {
```
Comment (mathpaquette) [inline packages/dashboard/src/components/layout/sidebar.tsx:646]: moved into projectDetailsMenu.tsx

File: packages/dashboard/src/components/layout/sidebar.tsx [hunk 4]
```diff
@@ -642,19 +499,3 @@ export const Sidebar: SidebarType = ({ open, onToggleSidebar }) => {
...
-  projectId: string;
-  projectColor?: string;
-  open: boolean;
-  onItemClick: () => void;
-  selectedItem: NavItemType;
-};
-type ProjectDetailsMenuType = React.FC<ProjectDetailsMenuProps>;
-
-type ProjectListMenuProps = {
```
Comment (mathpaquette) [inline packages/dashboard/src/components/layout/sidebar.tsx:655]: moved into projectListMenu.tsx

File: packages/dashboard/src/components/ui/searchField.tsx [hunk 5]
```diff
@@ -15,22 +15,33 @@ type SearchFieldProps = {
 const SearchField = ({ placeholder, onSearch, disabled }: SearchFieldProps) => {
```
Comment (mathpaquette) [inline packages/dashboard/src/components/ui/searchField.tsx:15]: @agoldis just improved the search making sure the parameter is being removed when no search value

File: packages/mongo/scripts/seed.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
+/**
```
Comment (mathpaquette) [inline packages/mongo/scripts/seed.ts:1]: @agoldis just a seed script, no impact on prod.
added tsconfig-paths/register in devDependencies

File: packages/dashboard/src/ciBuild/ciBuildSummary.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
+import { CardContent, Collapse, Grid, Typography } from '@mui/material';
+import {
+  getRunClaimedSpecsCount,
+  getRunDurationSeconds,
+  getRunOverallSpecsCount,
+  getRunTestsProgress,
+  getRunTestsProgressReducer,
+} from '@sorry-cypress/common';
```
Comment (mathpaquette) [inline packages/dashboard/src/ciBuild/ciBuildSummary.tsx:8]: reused all existing summary functions

**Comment by mathpaquette** [comment]
> @mathpaquette Quite a big PR, will try to review during the upcoming week

sorry, I thought upcoming weekS ... misread it :) I can wait.

**Review by agoldis — COMMENTED** [review]
Thanks @mathpaquette for your great work!

You have pointed a few possible improvements yourself , I will also add my 5 cents:

- "Recent Builds" or "Recent Runs" instead of "CI Builds" label?
- The sidebar navigation is confusing, may be the "home" view should be just the CI Build view
- Majority of users are not sharing the same CI build id across different projects. Thus, suggesting to make the item clickable if there's only one project
- Seems like there's no pagination or "Load more" for the list - any plans? 
- The query performance is a big bet :)

**Comment by mathpaquette** [comment]
> Thanks @mathpaquette for your great work!
Thank you @agoldis for maintaining this great alternative to Cypress :)
> 
> You have pointed a few possible improvements yourself , I will also add my 5 cents:
> 
> * "Recent Builds" or "Recent Runs" instead of "CI Builds" label?
Sure, let's try with Recent Builds

> * The sidebar navigation is confusing, may be the "home" view should be just the CI Build view
Maybe but at some point I'd like to use the Home to show some statistics like how many failed/flaky tests in last 24h, how many builds and so on. What do you think ? So basically for now, you're thinking to switching the landing page to CI Builds instead of projects?

> * Majority of users are not sharing the same CI build id across different projects. Thus, suggesting to make the item clickable if there's only one project
Yeah, absolutely. Let me do that.

> * Seems like there's no pagination or "Load more" for the list - any plans?
Yes, I wanted to add that once settled on the rest. Maybe a patch version ? :)

> * The query performance is a big bet :)
Honestly, I tested the performance alright on my data. I have 100k+ runs, they have lots of specs and tests. I'm very confident with it (also created the seed script to generate some to test this out). I was thinking to increase the batch to maybe to 2000 instead so it gives plenty of buffer to capture all runs. Since the query is just returning the first 20 groups, it means each CI Builds can have 100 projects before missing some. 

I'll update soon thanks !

**Comment by mathpaquette** [comment]
@agoldis could you please gimme an update on points 1-2. 051a3b9 should manage it.

**Comment by agoldis** [comment]
@mathpaquette  
> Maybe but at some point I'd like to use the Home to show some statistics like how many failed/flaky tests in last 24h, how many builds and so on. What do you think ? So basically for now, you're thinking to switching the landing page to CI Builds instead of projects?

I suggested that mainly to eliminate the confusion of the sidebar navigation. If "home" become the new CI Builds, then there would be no need to have the additional items in the side nav bar. As an alternative, probably "back" button would be sufficient. What do you think?

**Comment by agoldis** [comment]
@mathpaquette sorry for my sparse attention, please DM me on Slack if I am not responsive.

**Comment by mathpaquette** [comment]
> @mathpaquette
> 
> > Maybe but at some point I'd like to use the Home to show some statistics like how many failed/flaky tests in last 24h, how many builds and so on. What do you think ? So basically for now, you're thinking to switching the landing page to CI Builds instead of projects?
> 
> I suggested that mainly to eliminate the confusion of the sidebar navigation. If "home" become the new CI Builds, then there would be no need to have the additional items in the side nav bar. As an alternative, probably "back" button would be sufficient. What do you think?

Okay but how the sidemenu will look like when Home activated?

Home
   Recent Builds (highlighted)

**Comment by agoldis** [comment]
@mathpaquette Thanks a lot! Great addition to the current experience 💥🎉🔥 

I have applied my changes to the navbar and merged. Let's see if people need pagination / load more the the recent runs list.

**Comment by mathpaquette** [comment]
@agoldis i'll implement pagination. Thanks for your support !

**Comment by mathpaquette** [comment]
image

**Pull request merged**